File: backend/src/controllers/etaController.js
const pool = require('../config/database');
const redisClient = require('../config/redis');
const { getDistance } = require('geolib');

// Calculate ETA for a stop
const getStopETA = async (req, res) => {
  try {
    const { id } = req.params; // stop_id
    const { trip_id } = req.query;

    // Get stop location
    const stopResult = await pool.query('SELECT * FROM bus_stops WHERE id = $1', [id]);
    if (stopResult.rows.length === 0) {
      return res.status(404).json({ error: 'Stop not found' });
    }
    const stop = stopResult.rows[0];

    // Get trip details
    const tripResult = await pool.query('SELECT * FROM trips WHERE id = $1', [trip_id]);
    if (tripResult.rows.length === 0) {
      return res.status(404).json({ error: 'Trip not found' });
    }
    const trip = tripResult.rows[0];

    // Get current bus location from Redis
    const locationData = await redisClient.get(`bus:location:${trip.bus_id}`);
    if (!locationData) {
      return res.status(404).json({ error: 'Bus location not available' });
    }

    const location = JSON.parse(locationData);
    
    // Calculate distance in meters
    const distance = getDistance(
      { latitude: location.latitude, longitude: location.longitude },
      { latitude: stop.latitude, longitude: stop.longitude }
    );

    // Calculate ETA (assuming average speed of 30 km/h in city)
    const avgSpeed = location.speed > 0 ? location.speed : 30;
    const etaMinutes = Math.round((distance / 1000) / (avgSpeed / 60));

    res.json({
      stop_id: id,
      stop_name: stop.stop_name,
      distance_meters: distance,
      eta_minutes: etaMinutes,
      current_speed: location.speed,
    });
  } catch (error) {
    console.error('ETA calculation error:', error);
    res.status(500).json({ error: 'Failed to calculate ETA' });
  }
};

module.exports = { getStopETA };

File: backend-python/src/controllers/eta_controller.py
import json
from fastapi import APIRouter, HTTPException, Query
from src.config.database import get_pool
from src.config.redis import get_redis
from src.utils.distance import calculate_distance, calculate_eta

router = APIRouter(prefix="/stops", tags=["eta"])


@router.get("/{stop_id}/eta")
async def get_stop_eta(stop_id: int, trip_id: int = Query(...)):
    pool = await get_pool()
    redis = await get_redis()

    stop = await pool.fetchrow("SELECT * FROM bus_stops WHERE id = $1", stop_id)
    if not stop:
        raise HTTPException(status_code=404, detail="Stop not found")

    trip = await pool.fetchrow("SELECT * FROM trips WHERE id = $1", trip_id)
    if not trip:
        raise HTTPException(status_code=404, detail="Trip not found")

    cached = await redis.get(f"bus:location:{trip['bus_id']}")
    if not cached:
        raise HTTPException(status_code=404, detail="Bus location not available")

    location = json.loads(cached)
    distance = calculate_distance(
        {"lat": location["latitude"], "lng": location["longitude"]},
        {"lat": float(stop["latitude"]), "lng": float(stop["longitude"])},
    )
    speed = location.get("speed", 0)
    eta = calculate_eta(distance, speed)

    return {
        "stop_id": stop_id,
        "stop_name": stop["stop_name"],
        "distance_meters": round(distance, 2),
        "eta_minutes": eta,
        "current_speed": speed,
    }
